### PR TITLE
Bsodenkamp allocation identities refactor

### PR DESCRIPTION
**Pull request by bensodenkamp**
# Test Plan

bsodenkamp@Benjamins-MacBook-Pro cred % cat config/grain.json 
{
  "immediatePerWeek": 1000,
  "balancedPerWeek": 5000,
  "recentPerWeek": 19000,
  "recentWeeklyDecayRate": 0.16,
  "maxSimultaneousDistributions": 100
}
bsodenkamp@Benjamins-MacBook-Pro cred % pwd
/Users/bsodenkamp/projects/sourcecred.com/cred
bsodenkamp@Benjamins-MacBook-Pro cred % scdev grain -s -f > /tmp/scdevgrain                    
bsodenkamp@Benjamins-MacBook-Pro cred % scmain grain -s -f > /tmp/scmaingrain
bsodenkamp@Benjamins-MacBook-Pro cred % diff /tmp/scdevgrain /tmp/scmaingrain 
bsodenkamp@Benjamins-MacBook-Pro cred %

### CODE REVIEW

File: packages/sourcecred/src/core/ledger/grainAllocation.test.js [hunk 1]
```diff
@@ -61,35 +53,20 @@ describe("core/ledger/grainAllocation", () => {
...
-        expect(thunk).toThrowError("cred is zero");
-      });
-      it("errors if there's NaN or Infinity in Cred", () => {
-        const thunk = () =>
-          computeAllocation(immediate(5), [aid(0, [NaN])], credGrainView, 0);
-        expect(thunk).toThrowError("invalid cred");
-      });
-      it("errors if there's inconsistent Cred lengths", () => {
-        const i1 = aid(0, [1]);
-        const i2 = aid(0, [1, 2]);
```
Comment (bensodenkamp) [inline packages/sourcecred/src/core/ledger/grainAllocation.test.js:85]: So I removed these checks, and added a validation routine in the CredGrainView to do some of the same checks, however those checks are not tested. To test some of we would have to create a a credGrainView with no cred, or with mismatched interval data for participants, and I'm not sure how to go about that. Thoughts?

**Review by blueridger — DISMISSED** [review]

File: packages/sourcecred/src/core/ledger/grainAllocation.test.js (hunk 1 above)
Comment (blueridger) [reply packages/sourcecred/src/core/ledger/grainAllocation.test.js:85]: You can use the CredGrainView.fromJSON function to construct it manually with desired test data.

**Comment by bensodenkamp** [comment]
Added a bunch of tests for malformed credGrainView objects. Committed and manual tested.

File: packages/sourcecred/src/core/credGrainView.js [hunk 2]
```diff
@@ -131,20 +131,70 @@ export class CredGrainView {
...
+      if (p.credPerInterval.length !== this.intervals().length) {
+        throw new Error(`participant cred per interval length mismatch`);
+      }
+    });
+
+    this.activeParticipants().map((p) => {
+      if (p.grainEarnedPerInterval.length !== this.intervals().length) {
+        throw new Error(`participant grain per interval length mismatch`);
+      }
+    });
```
Comment (blueridger) [inline packages/sourcecred/src/core/credGrainView.js:156]: It'd be more efficient and concise if you placed all checks into the same loops. Not sure why the same loops are being duplicated throughout this function.

File: packages/sourcecred/src/core/credGrainView.test.js [hunk 3]
```diff
@@ -499,4 +499,201 @@ describe("core/credGrainView", () => {
       ).toThrow("The graph is missing account");
     });
   });
+
+  describe("validation tests", () => {
+    const credGrainViewNoCred = CredGrainView.fromJSON(
+      JSON.parse(
+        '{"participants":[{"active":true,"identity":{"id":"YVZhbGlkVXVpZEF0TGFzdA", "subtype":"USER",\
```
Comment (blueridger) [inline packages/sourcecred/src/core/credGrainView.test.js:506]: This works. An alternative strategy could've been to initialize a valid data structure in a beforeEach, and then modify it to be invalid for each test case. Parsing from string is kinda beneficial to smush it into fewer lines, but you could also have just initialized the "JSON" as a javascript hash and that would keep Flow typing stricter.

Comment (bensodenkamp) [reply packages/sourcecred/src/core/credGrainView.test.js:506]: Noted. I think my motivation was just to be able to see all the data at once.

File: packages/sourcecred/src/core/credGrainView.js (hunk 2 above)
Comment (bensodenkamp) [reply packages/sourcecred/src/core/credGrainView.js]: fixed

File: packages/sourcecred/src/core/credGrainView.js [hunk 4]
```diff
@@ -141,60 +145,44 @@ export class CredGrainView {
...
-    this.activeParticipants().map((p) => {
       if (!G.eq(G.sum(p.grainEarnedPerInterval), p.grainEarned)) {
         throw new Error(
           `participant grain per interval mismatched with participant grain total:`
         );
       }
-    });
 
-    this.activeParticipants().map((p) => {
       p.grainEarnedPerInterval.map((g) => {
```
Comment (blueridger) [inline packages/sourcecred/src/core/credGrainView.js:169]: map -> forEach

File: packages/sourcecred/src/core/credGrainView.js [hunk 5]
```diff
@@ -141,60 +145,44 @@ export class CredGrainView {
...
         if (g < G.ZERO) {
           throw new Error(`negative grain paid in interval data`);
         }
       });
-    });
 
-    this.activeParticipants().map((p) => {
       p.credPerInterval.map((c) => {
```
Comment (blueridger) [inline packages/sourcecred/src/core/credGrainView.js:175]: map -> forEach
also can be deduped

**Review by blueridger — APPROVED** [review]
so close! just missed 2
you can merge on this approval once fixed

**Pull request merged**